sympy/core/add.py

```diff
@@ -862,10 +862,33 @@ def _eval_as_leading_term(self, x):
 
         infinite = [t for t in expr.args if t.is_infinite]
 
-        expr = expr.func(*[t.as_leading_term(x) for t in expr.args]).removeO()
-        if not expr:
-            # simple leading term analysis gave us 0 but we have to send
-            # back a term, so compute the leading term (via series)
+        compute = False
+        leading_terms = [t.as_leading_term(x) for t in expr.args]
+        new_expr = expr.func(*leading_terms).removeO()
+
+        if not new_expr:
+            compute = True
+        elif new_expr is not S.NaN:
+            if new_expr.is_Add:
+                final_leading_terms = [t for t in new_expr.args]
+            else:
+                final_leading_terms = [new_expr]
+
+            if len(expr.args) != len(final_leading_terms):
```

Comment:
Can code below this be merged with #L875 (``final_leading_terms = [new_expr]``)

Reply:
I couldn't get you. @leosartaj . Could you please explain?

Reply:
Ping @leosartaj